Modernize pacemaker-schedulerd

clumens opened the pull request:
This patch set does three things:

* Convert pacemaker-schedulerd to use glib for its command line handling.
* Rearrange how pacemaker-schedulerd uses formatted output.
* Convert pacemaker-schedulerd to use the new IPC API mechanism.

clumens commented on daemons/schedulerd/pacemaker-schedulerd.c:
```diff
@@ -296,68 +301,57 @@ struct qb_ipcs_service_handlers ipc_callbacks = {
+        pcmk__str_eq("metadata", options.remainder[0], pcmk__str_casei)) {
         pe_metadata();
         goto done;
     }
```
Fixed.

clumens commented on lib/common/ipc_schedulerd.c:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    private = api->api_data;
+    /* See comments in ipc_pacemakerd.c and ipc_controld.c. */
```
Fixed.  I had wondered if I'd need to take care of the proxy side of things.  Good thing to know I don't.

clumens commented on lib/common/ipc_schedulerd.c:
```diff
@@ -0,0 +1,196 @@
+    CRM_ASSERT(private != NULL);
+
+    *cmd = create_request(task, xml_data, NULL, CRM_SYSTEM_PENGINE,
+                          ipc_name?ipc_name:((crm_system_name? crm_system_name : "client")),
```
If I can get the function arguments straightened out (see your comments below about the request XML), then this whole function is almost identical to do_pacemakerd_api_call.  I should probably just combine those into a common private function somewhere.

clumens commented on lib/common/ipc_schedulerd.c:
```diff
@@ -0,0 +1,196 @@
+
+static int
+do_schedulerd_api_call(pcmk_ipc_api_t *api, const char *ipc_name, const char *task,
+                       xmlNode *xml_data, xmlNode **cmd)
```
How would I go about getting the reference to pass to controld_expect_sched_reply() then?  I could turn it into an out parameter from do_schedulerd_api_call and that whole call chain, but hopefully there's a better way.

clumens commented on daemons/schedulerd/pacemaker-schedulerd.c:
```diff
@@ -325,30 +325,42 @@ main(int argc, char **argv)
         goto done;
     }
 
+    rc = pcmk__output_new(&out, args->output_ty, args->output_dest, argv);
```
If we're going to do that, we don't necessarily need the patch that allows args when spawning subdaemons (ad03849), though code-wise I do prefer that to the static array dance we are currently doing there.

clumens commented on daemons/controld/controld_schedulerd.c:
```diff
@@ -464,9 +477,7 @@ do_pe_invoke_callback(xmlNode * msg, int call_id, int rc, xmlNode * output, void
-    cmd = create_request(CRM_OP_PECALC, output, NULL, CRM_SYSTEM_PENGINE, CRM_SYSTEM_DC, NULL);
-
-    rc = pe_subsystem_send(cmd);
+    rc = pe_subsystem_send(output, &cmd);
```
Added a patch to change parameters to controld_expect_sched_reply.

kgaillot commented on daemons/controld/controld_schedulerd.c:
```diff
@@ -110,67 +111,94 @@ pe_ipc_destroy(gpointer user_data)
+    rc = pcmk_connect_ipc(schedulerd_api, pcmk_ipc_dispatch_main);
+    if (rc != pcmk_rc_ok) {
+        crm_err("Error connecting to the scheduler: %s", pcmk_rc_str(rc));
+        return false;
```
I'm thinking we should free and null schedulerd_api here

clumens replied:
Fixed.

kgaillot commented on daemons/controld/controld_schedulerd.c:
```diff
@@ -182,19 +210,12 @@ pe_subsystem_new(void)
+        return pcmk_schedulerd_api_graph(schedulerd_api, CRM_SYSTEM_DC, xml_data, ref);
     }
+
     return -ENOTCONN;
```
At this point I don't think this is worth a function anymore

kgaillot commented on lib/common/ipc_schedulerd.c:
```diff
@@ -0,0 +1,178 @@
+    xmlNode *cmd = NULL;
+    int rc;
+
+    CRM_CHECK(api != NULL, return -EINVAL);
```
Since this (actually the caller) is public API, I think a regular if is better than CRM_CHECK()

Also + not -

clumens commented on daemons/controld/controld_schedulerd.c:
```diff
@@ -110,67 +111,94 @@ pe_ipc_destroy(gpointer user_data)
 
     controld_clear_fsa_input_flags(R_PE_CONNECTED);
-    pe_subsystem = NULL;
+    schedulerd_api = NULL;
```
Fixed.

clumens commented on lib/common/ipc_schedulerd.c:
```diff
@@ -0,0 +1,178 @@
+
+    // Parse useful info from reply
+    msg_data = get_message_xml(reply, F_CRM_DATA);
+    crm_element_value_ll(msg_data, XML_ATTR_TSTAMP, &value_ll);
```
Fixed.

clumens commented:
Fixed authorship information and fixed a bug in the last patch that coverity caught.  I'm still running CTSlab against this, and I still need to get to valgrind.  So it's not ready to merge quite yet, but I wanted a backup copy of this to exist somewhere.

clumens commented on include/crm/common/ipc_schedulerd.h:
```diff
@@ -0,0 +1,54 @@
+    } data;
+} pcmk_schedulerd_api_reply_t;
+
+int pcmk_schedulerd_api_graph(pcmk_ipc_api_t *api, xmlNode *xml_data, char **ref);
```
Fixed.

clumens commented:
Freeing the API in pe_subsystem_new seems to have reduced the number of leaks, but there are still a few outstanding.  Freeing the API in pe_ipc_destroy leads to a lot more memory problems, so that probably needs to be something more clever.  Running CTSlab ten times results in three leaks of approximately 140 bytes each because of this.

There's also this:
```
==46559== 224 bytes in 4 blocks are definitely lost in loss record 118 of 140
==46559==    at 0x4C34F0B: malloc (vg_replace_malloc.c:307)
==46559==    by 0x884680D: strdup (in /usr/lib64/libc-2.28.so)
==46559==    by 0x574F6ED: lrmd__set_result (lrmd_client.c:2375)
==46559==    by 0x121B84: fail_pending_op (controld_execd_state.c:79)
==46559==    by 0x817768F: ??? (in /usr/lib64/libglib-2.0.so.0.5600.4)
==46559==    by 0x122F32: lrm_state_disconnect_only (controld_execd_state.c:333)
==46559==    by 0x123073: lrm_state_disconnect (controld_execd_state.c:345)
==46559==    by 0x137BC6: monitor_timeout_cb (controld_remote_ra.c:493)
==46559==    by 0x818A240: ??? (in /usr/lib64/libglib-2.0.so.0.5600.4)
==46559==    by 0x818977C: g_main_context_dispatch (in /usr/lib64/libglib-2.0.so.0.5600.4)
==46559==    by 0x8189B47: ??? (in /usr/lib64/libglib-2.0.so.0.5600.4)
==46559==    by 0x8189E71: g_main_loop_run (in /usr/lib64/libglib-2.0.so.0.5600.4)
==46559==    by 0x110FD2: crmd_init (pacemaker-controld.c:160)
==46559==    by 0x110DED: main (pacemaker-controld.c:130)
```
which I think can be solved by just freeing the event.exit_reason at the end of fail_pending_op.  I think when process_lrm_event is done, we're done with the whole event variable.  But I could use a double check of that.

Also, the Reattach CTSlab test seems to somewhat consistently fail:
```
Oct 29 15:05:04 Running test RemoteStonithd         (ctslab2)     [  6]
Oct 29 15:12:33 Running test Reattach               (ctslab1)     [  7]
Oct 29 15:18:16 ResourceActivity: Oct 29 15:15:49 ctslab1 pacemaker-controld[90980]:  notice: Result of start operation for Fencing on ctslab1: ok
Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:10 ctslab1 pacemaker-controld[90980]:  notice: Result of stop operation for lsb-dummy on ctslab1: ok
Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:10 ctslab1 pacemaker-controld[90980]:  notice: Result of stop operation for r192.168.122.104 on ctslab1: ok
Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:21 ctslab1 pacemaker-controld[90980]:  notice: Result of stop operation for petulant on ctslab1: ok
Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:32 ctslab1 pacemaker-controld[90980]:  notice: Result of start operation for petulant on ctslab1: ok
Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:32 ctslab1 pacemaker-controld[90980]:  notice: Result of start operation for r192.168.122.104 on ctslab1: ok
Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:32 ctslab1 pacemaker-controld[90980]:  notice: Result of start operation for lsb-dummy on ctslab1: ok
Oct 29 15:18:31 ResourceActivity: Oct 29 15:16:03 ctslab2 pacemaker-controld[6044]:  notice: Result of start operation for FencingFail on ctslab2: ok
Oct 29 15:18:31 ResourceActivity: Oct 29 15:16:07 ctslab2 pacemaker-controld[6044]:  notice: Result of stop operation for stateful-1 on ctslab2: ok
Oct 29 15:18:35 ResourceActivity: Oct 29 15:16:21 ctslab2 pacemaker-controld[6044]:  notice: Result of start operation for stateful-1 on ctslab2: ok
Oct 29 15:18:56 Test Reattach                       FAILED: Resources stopped or started during cluster restart
```

kgaillot commented:
> Freeing the API in pe_subsystem_new seems to have reduced the number of leaks, but there are still a few outstanding. Freeing the API in pe_ipc_destroy leads to a lot more memory problems, so that probably needs to be something more clever. Running CTSlab ten times results in three leaks of approximately 140 bytes each because of this.

Oh of course, pe_ipc_destroy() is called by the IPC callback, when it returns the IPC code resumes but the object is now dead. We may need to switch up the model a bit, instead of creating/destroying the object repeatedly, create it once if null, and free it once at shutdown. otherwise just connect/disconnect. that's a better approach anyway, the old model didn't support it well

pe_ipc_destroy() simply wouldn't null the object. everywhere that currently checks whether the object is non-null would instead have to check whether it's connected. pe_subsystem_new() would not free/null any existing object but just reconnect if needed

> 
> There's also this:
> 
> ```
> ==46559== 224 bytes in 4 blocks are definitely lost in loss record 118 of 140
> ==46559==    at 0x4C34F0B: malloc (vg_replace_malloc.c:307)
> ==46559==    by 0x884680D: strdup (in /usr/lib64/libc-2.28.so)
> ==46559==    by 0x574F6ED: lrmd__set_result (lrmd_client.c:2375)
> ==46559==    by 0x121B84: fail_pending_op (controld_execd_state.c:79)
> ==46559==    by 0x817768F: ??? (in /usr/lib64/libglib-2.0.so.0.5600.4)
> ==46559==    by 0x122F32: lrm_state_disconnect_only (controld_execd_state.c:333)
> ==46559==    by 0x123073: lrm_state_disconnect (controld_execd_state.c:345)
> ==46559==    by 0x137BC6: monitor_timeout_cb (controld_remote_ra.c:493)
> ==46559==    by 0x818A240: ??? (in /usr/lib64/libglib-2.0.so.0.5600.4)
> ==46559==    by 0x818977C: g_main_context_dispatch (in /usr/lib64/libglib-2.0.so.0.5600.4)
> ==46559==    by 0x8189B47: ??? (in /usr/lib64/libglib-2.0.so.0.5600.4)
> ==46559==    by 0x8189E71: g_main_loop_run (in /usr/lib64/libglib-2.0.so.0.5600.4)
> ==46559==    by 0x110FD2: crmd_init (pacemaker-controld.c:160)
> ==46559==    by 0x110DED: main (pacemaker-controld.c:130)
> ```
> 
> which I think can be solved by just freeing the event.exit_reason at the end of fail_pending_op. I think when process_lrm_event is done, we're done with the whole event variable. But I could use a double check of that.
>

Good catch -- that's unrelated to your changes, recently introduced. Call lrmd__reset_result() at the end there.

> Also, the Reattach CTSlab test seems to somewhat consistently fail:
> 
> ```
> Oct 29 15:05:04 Running test RemoteStonithd         (ctslab2)     [  6]
> Oct 29 15:12:33 Running test Reattach               (ctslab1)     [  7]
> Oct 29 15:18:16 ResourceActivity: Oct 29 15:15:49 ctslab1 pacemaker-controld[90980]:  notice: Result of start operation for Fencing on ctslab1: ok
> Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:10 ctslab1 pacemaker-controld[90980]:  notice: Result of stop operation for lsb-dummy on ctslab1: ok
> Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:10 ctslab1 pacemaker-controld[90980]:  notice: Result of stop operation for r192.168.122.104 on ctslab1: ok
> Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:21 ctslab1 pacemaker-controld[90980]:  notice: Result of stop operation for petulant on ctslab1: ok
> Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:32 ctslab1 pacemaker-controld[90980]:  notice: Result of start operation for petulant on ctslab1: ok
> Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:32 ctslab1 pacemaker-controld[90980]:  notice: Result of start operation for r192.168.122.104 on ctslab1: ok
> Oct 29 15:18:20 ResourceActivity: Oct 29 15:16:32 ctslab1 pacemaker-controld[90980]:  notice: Result of start operation for lsb-dummy on ctslab1: ok
> Oct 29 15:18:31 ResourceActivity: Oct 29 15:16:03 ctslab2 pacemaker-controld[6044]:  notice: Result of start operation for FencingFail on ctslab2: ok
> Oct 29 15:18:31 ResourceActivity: Oct 29 15:16:07 ctslab2 pacemaker-controld[6044]:  notice: Result of stop operation for stateful-1 on ctslab2: ok
> Oct 29 15:18:35 ResourceActivity: Oct 29 15:16:21 ctslab2 pacemaker-controld[6044]:  notice: Result of start operation for stateful-1 on ctslab2: ok
> Oct 29 15:18:56 Test Reattach                       FAILED: Resources stopped or started during cluster restart
> ```

I haven't seen that, it must be an issue with your test environment or related to your changes. You can see what Reattach does in cts/lab/CTStests.py, follow it step by step through your CTS log and the cluster logs to see what went awry. FYI I have an alias that helps remove a lot of the useless crud from a CTS log for easier browsing:

    alias slim-cts-log="grep -v 'debug: cmd: async: .*journalctl -q' cts.log \
        | grep -v '\(ClusterStable\|fast-start\): Waiting:' \
        | grep -v '\(ClusterStable\|fast-start\): Got .* lines' \
        | grep -v 'debug: cmd: pid .* returned 0' \
        > cts.log.slim"

clumens commented:
All the "Result of %s operation for %s" messages were added recently by 9a197c732e, so I think those just need to be ignored in the test (or at least, as long as they are ok they should be ignored).  I'll do that and we'll take it from there.

clumens commented:
Added the memory management as a patch on top so we can talk about it separately, and also added a patch to ignore the ResourceActivity message.  I think this could probably be moved to somewhere in cts/lab/patterns.py, but I haven't sorted all that code out yet.

kgaillot commented:
> All the "Result of %s operation for %s" messages were added recently by [9a197c7](https://github.com/ClusterLabs/pacemaker/commit/9a197c732ea17660091976a5366de9fed414aad1), so I think those just need to be ignored in the test (or at least, as long as they are ok they should be ignored). I'll do that and we'll take it from there.

As discussed elsewhere, these are real problems and should not be ignored in the lab code. If it's run under valgrind, all lab messages can be ignored.

clumens commented:
> > All the "Result of %s operation for %s" messages were added recently by [9a197c7](https://github.com/ClusterLabs/pacemaker/commit/9a197c732ea17660091976a5366de9fed414aad1), so I think those just need to be ignored in the test (or at least, as long as they are ok they should be ignored). I'll do that and we'll take it from there.
> 
> As discussed elsewhere, these are real problems and should not be ignored in the lab code. If it's run under valgrind, all lab messages can be ignored.

Yeah, that must have been happening under valgrind.  I did another run of CTSlab this morning with a bunch of Reattach tests run and never saw any of these error messages.  I've dropped the patch, too.

kgaillot reviewed (commented):
Getting there! Can you submit the lrmd__reset_result() commit in a separate PR against 2.1? I want to make sure we get it in the release.

clumens commented on lib/common/ipc_schedulerd.c:
```diff
@@ -143,7 +143,7 @@ do_schedulerd_api_call(pcmk_ipc_api_t *api, const char *task, xmlNode *cib, char
     int rc;
 
-    if (api == NULL) {
+    if (api == NULL || !pcmk_ipc_is_connected(api)) {
```
Fixed.

clumens commented on daemons/controld/controld_schedulerd.c:
```diff
@@ -32,13 +34,11 @@ void
-        controld_clear_fsa_input_flags(R_PE_CONNECTED);
-    }
+    pcmk_disconnect_ipc(schedulerd_api);
+    pe_ipc_destroy();
```
Renamed these two and pe_subsystem_new as well (which was also a static function).

kgaillot commented on daemons/controld/controld_schedulerd.c:
```diff
@@ -187,25 +187,25 @@ pe_subsystem_new(void)
+        if (rc != pcmk_rc_ok) {
+            crm_err("Error connecting to the scheduler: %s", pcmk_rc_str(rc));
+            pcmk_free_ipc_api(schedulerd_api);
+            schedulerd_api = NULL;
```
Just noticed we don't need to free and null the api object here, just leave it disconnected

Also, pcmk_connect_ipc() will return ok if it's already connected, so we really don't even need to check that at all

clumens replied:
Fixed.  Nice to remove some unneeded checks and simplify the code.

kgaillot commented on lib/common/ipc_schedulerd.c:
```diff
@@ -0,0 +1,177 @@
+    int rc;
+
+    if (!pcmk_ipc_is_connected(api)) {
+        return EINVAL;
```
This should be ENOTCONN. Then we don't need to check for connected separately before calling api_graph()

clumens replied:
Fixed and merged with d9d2240, since that patch is was introduces using pcmk_ipc_is_connected in all these places.

kgaillot commented on daemons/controld/controld_schedulerd.c:
```diff
@@ -31,11 +31,11 @@ static pcmk_ipc_api_t *schedulerd_api = NULL;
  */
 void
-pe_subsystem_free(void)
+controld_shutdown_ipc(void)
```
The name should indicate that this is specifically for the scheduler IPC -- the controller has its own IPC, and also connects to executor and fencer IPC

clumens replied:
Fixed with some better names, hopefully.

kgaillot reviewed (commented):
Looking good, just minor tweaks

kgaillot commented on daemons/schedulerd/pacemaker-schedulerd.c:
```diff
@@ -321,19 +323,33 @@ int
+    }
+
+    pe__register_messages(out);
+    pcmk__register_lib_messages(out);
```
Not necessary with this PR, but I don't think we need to register any messages for the main output object since it's only used with version/help. That could change when we implement the metadata output though

kgaillot commented on daemons/controld/controld_schedulerd.c:
```diff
@@ -209,15 +217,16 @@ do_pe_control(long long action,
     if (action & A_PE_STOP) {
-        pe_subsystem_free();
+        controld_clear_fsa_input_flags(R_PE_REQUIRED);
+        handle_disconnect();
```
We have to actually disconnect here, before handle_disconnect()

clumens replied:
Oh, of course.  Fixed.

kgaillot reviewed (commented):
One final change and it's good :)

kgaillot commented:
Another step forward :)
